#159987402 implement password reset

tomiadebanjo opened the pull request:
#### What does this PR do?
Implement password reset

#### Description of Task to be completed?
- Install sendgridAPI
- create endpoints for users to request for a password reset.
- setup functionality to allow users reset their password.

#### How should this be manually tested?
1. Clone the repository.
2. Checkout the `ft-user-password-reset-159987402` branch.
3. Run npm install on your local machine.
4. Run the db migrations using `npm run migrate`.
5. Start the server using `npm run start_dev`.
5. You can test the following endpoints using postman
`localhost:8000/api/v1/users/api/auth/reset-password/` 
`localhost:8000/api/v1/users/api/auth/reset-password/:token`

#### What are the relevant pivotal tracker stories?
[ #159987402](https://www.pivotaltracker.com/story/show/159987402)

iAmao commented on server/controllers/userController.js:
```diff
@@ -66,6 +75,58 @@ const userController = {
+      .findOne({
+        where: { email: req.body.email }
+      })
+      .then((user) => {
```
No catch block for this. If something happens users will never get a response

tomiadebanjo replied:
Fixed

iAmao commented on server/controllers/userController.js:
```diff
@@ -66,6 +75,58 @@ const userController = {
+  reset: (req, res) => {
+    Users.findOne({
+      where: { id: req.currentUser.id }
+    }).then((user) => {
```
No catch block for this. If something happens users will never get a response

tomiadebanjo replied:
Fixed

iAmao commented on server/controllers/userController.js:
```diff
@@ -66,6 +75,58 @@ const userController = {
+        password: bcrypt.hashSync(req.body.password, 8)
+      }, {
+        where: { id: req.currentUser.id }
+      }).then(() => res.status(200).jsend.success({
```
No catch block for this. If something happens users will never get a response

tomiadebanjo replied:
I've added catch blocks for all the then's.

iAmao commented on server/helpers/generateToken.js:
```diff
@@ -7,12 +7,12 @@ const secret = process.env.SECRET;
  */
-const generateToken = (id, time) => {
-  const rawToken = jwt.sign({ id }, secret, { expiresIn: time });
+const generateToken = (time, ...args) => {
```
You could just use an object instead of a spread operator

iAmao reviewed (changes requested):
Nice, but always handle edge cases and errors, keep function as simple as possible.

tomiadebanjo committed:
feat(authenication): implement password reset

- user receives link to reset password via email
- user can reset password with the link received via email
- user can only use the link received via email once

[Delivers #159987402]

tomiadebanjo committed:
Implement changes from TTL review
